### ISSUE TITLE
z/OS: when using Node.js for z/OS, how to set the tempDir for '.com_ibm_tools_hc' (semaphore resource directory)

### ISSUE THREAD

**Issue by liuskhk1** [RuntimeTools/omr-agentcore#96]
z/OS specific:
When running node.js app under z/OS with appmetrics/appmetrics-dash, it will try to write some files under the /tmp folder, like:
```
[18.47.19 S0055777 ICH408I USER(ASU0C24K) GROUP(GSTRIN03) NAME(CICS, DEV U0 ) 055
055 /tmp/.com_ibm_tools_hc/common_env_notifier
055 CL(DIRACC ) FID(00000000000000000000083900000000)
055 INSUFFICIENT AUTHORITY TO OPEN
055 ACCESS INTENT(-W-) ACCESS ALLOWED(OTHER R-X)
055 EFFECTIVE UID(0000007403) EFFECTIVE GID(0000000020) ]
```
Have already tried to set z/OS temp dir to other path and restart the node.js app, but it still tried to write to /tmp
```
export TMPDIR=/my/project/tmp
```
Checked the code from:
https://github.com/RuntimeTools/omr-agentcore/blob/master/src/ibmras/common/port/zos/Thread.cpp

```
180 bool getIPCKey(key_t* handle, std::string name) {
181    // get the semaphore temp dir set in the loader
182    std::string tempDir = ibmras::monitoring::agent::Agent::getInstance()->getProperty("platform_tempdir");
183    if ("" == tempDir) {
184        IBMRAS_DEBUG(debug, "Defaulting tempDir to /tmp");
185        tempDir = "/tmp";
186    }
```
Looks like it will call `getProperty("platform_tempdir")`.  How to set the path for `platform_tempdir`? Through a property file? Or by another environment variable?
Please kindly advise. Many thanks.

**Comment by mattcolegate** [comment]
Hi @liuskhk1, you need to provide it as a parameter on the command line when starting your program e.g. if you ran your program with `java -Xhealthcenter MyProgram` then you would change the command to `java -Xhealthcenter -Dplatform_tempdir=/my/project/tmp MyProgram`.
Please close this issue if this solves your problem.

**Comment by liuskhk1** [comment]
Hi @mattcolegate we are using Node.js for z/OS instead of Java. (remarks: we don't use healthcenter as it is not support on z/OS). Is there any equvalent parameter for Node.js for z/OS?

Here is my sample coding of using appmetrics/appmetrics-dash:
```
var express = require('express');
var app = express();
var appmetrics = require('appmetrics');
var monitoring = appmetrics.monitor();
var dash = require('appmetrics-dash').attach();
app.get('/', function (req, res) {
 res.send('Hello World!');
});
app.listen(30000, function () {
 console.log('Example app listening on port 30000!');
});
```
Thank you very much.

**Comment by mattcolegate** [comment]
For Node.js you can specify the variable in `appmetrics.properties` or set it at runtime using a call to `configure`. The section describing how to do so in the readme is here: https://www.npmjs.com/package/appmetrics#configuring-node-application-metrics

**Comment by liuskhk1** [comment]
Hi Matt, thanks for your reply. I've tried the following:

1) add `platform_tempdir=/u/ibmuser/tmp` in appmetrics.properties file.
(Remarks: The appmetrics readme says Health Center Eclipse IDE client is Not supported on z/OS, not sure if it will read the appmetrics.properties file)

2) use appmetrics.configure to set the variable, e.g.:
```
var appmetrics = require('appmetrics');
appmetrics.configure({'platform_tempdir':'/u/ibmuser/tmp'});
```

3) set the env var:
```
export TMP=/u/ibmuser/tmp
export TMPDIR=/u/ibmuser/tmp
export PLATFORM_TEMPDIR=/u/ibmuser/tmp
```
After all, the problem still persists.

**Comment by mattcolegate** [comment]
Drat. I'll try and get hold of a proper z/OS test system and debug. Should take a few days with my current workload.

**Comment by mattcolegate** [comment]
Hi, just to say that I am still working on the problem but I've not found a solution yet. I have a test rig up and running. I hope to have good news soon. I will update again at the end of the week.

**Comment by mattcolegate** [comment]
Hi @liuskhk1, can you try the following snippet of code for me on your z/OS system? This outputs as expected on my Windows box:
```js
var appmetrics = require('appmetrics');
appmetrics.configure({'platform_tempdir': 'this should work'});
var monitoring = appmetrics.monitor();
console.log(monitoring.agent.getOption('platform_tempdir'));
```
If you could also run it with the following option in `appmetrics.properties` this would hopefully allow us to see whether we are defaulting to `/tmp`:
```js
com.ibm.diagnostics.healthcenter.logging.Port=debug
```

**Comment by mattcolegate** [comment]
The thought occurs that there might be some ASCII/EBCDIC issue going on with the setting of the variable - can the above also be tried with `platform_tempdir` being translated to it's ASCII counterpart?

**Comment by liuskhk1** [comment]
Hello @mattcolegate, my previous zD&T testing environment was no longer available due to the file corruption of the zD&T VM image. I had to rebuild my zD&T environment. In the new environment, I have tried to add `com.ibm.diagnostics.healthcenter.logging.Port=debug` to `appmetrics.properties` and then run the following code:

```
var appmetrics = require('appmetrics');
appmetrics.configure({'platform_tempdir': 'this should work'});
var monitoring = appmetrics.monitor();
console.log(monitoring.agent.getOption('platform_tempdir'));
```

The output is:
```
IBMUSER:/u/ibmuser/myapp: >node server.js
[Tue Apr 28 13:15:16 2020] com.ibm.diagnostics.healthcenter.loader INFO: Node Application Metrics 5.1.1.202002041855 (Agent Core 4.0.5)
this should work
IBMUSER:/u/ibmuser/myapp: >
```

In the above case, I did not see directory `.com_ibm_tools_hc` created on `/tmp`

After changing the code as following:

```
var appmetrics = require('appmetrics');
appmetrics.configure({'platform_tempdir': '/u/ibmuser/tmp'});
var monitoring = appmetrics.monitor();
console.log(monitoring.agent.getOption('platform_tempdir'));
```

This time I can see directory `.com_ibm_tools_hc` was created under `/u/ibmuser/tmp` and it was also displayed by console.log(). I don't why it did not work in my previous zD&T environment, but this time `appmetrics.configure({'platform_tempdir': '/u/ibmuser/tmp'});` works perfectly.
(Remarks: the server.js is encoded in IBM-1047.)

I will ask my client to test it in his environment, and I will update this issue after receiving my client's feedback.
I much appreciate your kind support and assistance. Thank you very much.

**Comment by liuskhk1** [comment]
Hi @mattcolegate, my client tried the following snippet of code:

```
var appmetrics = require('appmetrics');
appmetrics.configure({'platform_tempdir': '/u/user/local/m40031'});
var monitoring = appmetrics.monitor();
console.log(monitoring.agent.getOption('platform_tempdir'));
```

and replied as follow:
```
The STDOUT can show the path I setup in 'platform_tempdir and I can find the directory '.com_ibm_tools_hc/’ there.  It looks good.

However, I find the following message in STDERR.  It seems it will create a resourceRegistry.log file under the directory where my server.js (the first javascript file to start) located by not refer to the tempdir setup:
[Wed Apr 29 05:53:42 2020] com.ibm.diagnostics.healthcenter.loader INFO: Node Application Metrics 5.1.1.202003260107 (Agent Core 4.0.5)
log4js.fileAppender - Writing to file resourceRegistry.log, error happened  { Error: EACCES: permission denied, open '/u/user/local/m40031/sprint2/obsuat/resourceRegistry.log'
  errno: -111,
  code: 'EACCES',
  syscall: 'open',
  path: '/u/user/local/m40031/sprint2/obsuat/resourceRegistry.log' }

Since we are no sure what resourceRegistry.log is for, may I ask if we have multiple CICS bundles running the same server.js, and suppose they will refer to the same resourceRegistry.log file.  Will it cause any issue, e.g. data overridden.
```

I can find that 'resourceRegistry.log' is mentioned in .gitignore:
https://github.com/RuntimeTools/appmetrics/blob/master/.gitignore

Could you please further help and advise?
If a new ticket should be created to pursue the resourceRegistry.log issue, please let me know.
Thank you very much.

**Comment by mattcolegate** [comment]
Hi @liuskhk1,
resourseRegistry.log comes from `ibmapm-restclient`, which is a dependency of `ibmapm-embed`, which is a dependency of `appmetrics`.
https://github.com/IBM/node-ibmapm-restclient/blob/96bdad6ba089e77d823c42a569d50983ba6c2978/lib/plugins/logutil.js
I've no idea what it's used for - it seems to be 0 bytes on my system.
I can't seem to stop the file being created, but your client may be able to avoid this error by specifying the environment variable `KNJ_LOG_LEVEL=OFF`
Otherwise the matter should be taken up by raising an issue with ibmapm-restclient.

**Comment by mattcolegate** [comment]
Looks like someone beat you to it: https://github.com/IBM/node-ibmapm-restclient/issues/5

**Comment by liuskhk1** [comment]
Received client's feedback:
- - - - - - - - - - - - - - - - - 
Thanks for your information.
I have tried KNJ_LOG_LEVEL=OFF and the file will still be created.  I have created an empty file and chmod 777 so all bundle can update it.  As mentioned in github, the file keep as 0 bytes after doing transactions.  So it should be ok now.  Thanks. 
- - - - - - - - - - - - - - - - - 

Problem solved. This issue can be closed. Thanks @mattcolegate again for your help.
